File: frontend/src/game/upgrades/stat_modifiers/tenacity_2.ts
import { Upgrade, type UpgradeDef } from '../Upgrade'
import { UpgradeTargetID, UpgradeStatID, RarityID, UpgradeTypeID } from '../../data/ID'

export const Tenacity2Def: UpgradeDef = {
  id: "tenacity_2",
  name: "Tenacity",
  description: "Increased the bullet lifespan by .4 seconds",
  rarity: RarityID.Epic,
  upgradeType: UpgradeTypeID.StatModifier,
  cost: 20,
  targetClass: UpgradeTargetID.Bullet,
  fieldInTargetClass: UpgradeStatID.TimeLeft,
  value: 400,
  specificAttackType: "bullet",
  stackable: true,
  maxStacks: 3,
}

export class Tenacity2 extends Upgrade {}

File: backend/app/core/upgrades/stat_modifiers/tenacity_2.py
from ..upgrade_implementation import UpgradeImplementation, UpgradeAppliers


class Tenacity2(UpgradeImplementation):
    id = 'tenacity_2'
    name = 'Tenacity'
    description = 'Increased the bullet lifespan by .4 seconds'
    rarity = 'epic'
    type = 'stat_modifier'
    cost = 20

    def can_apply(self, state: dict) -> bool:
        return True

    def apply(self, appliers: UpgradeAppliers) -> None:
        value = 400
        appliers.modifiers.add_multiplier('bullet', 'timeLeft', value)

    def validate(self, recorded_stats: dict) -> bool:
        return True
